### PR TITLE
Admin command to check if sealed extents are sealed on replicas

### PR DESCRIPTION
**Pull request by KiranRG**
This adds an admin command that iterates through extents and checks to ensure that if the extent is sealed in metadata, that it is sealed on all of the replicas, as well.

### CODE REVIEW

File: clients/storehost/client.go [hunk 1]
```diff
@@ -67,3 +67,19 @@ func (s *StoreClientImpl) ReadMessages(req *store.ReadMessagesRequest) (*store.R
...
+
+// GetAddressFromTimestamp queries store for the address corresponding to the given timestamp
+func (s *StoreClientImpl) GetAddressFromTimestamp(req *store.GetAddressFromTimestampRequest) (*store.GetAddressFromTimestampResult_, error) {
+	ctx, cancel := tcthrift.NewContext(2 * time.Second)
+	defer cancel()
+
+	return s.client.GetAddressFromTimestamp(ctx, req)
+}
+
+// SealExtent queries store for the address corresponding to the given timestamp
```
Comment (venkat1109) [inline clients/storehost/client.go]: fix comment

File: tools/common/lib.go [hunk 2]
```diff
@@ -1079,3 +1080,148 @@ func getIfSetString(c *cli.Context, p string, setCount *int) (r *string) {
...
+	return client, nil
+}
+
+func (t *storeClientCache) close() {
+	for _, client := range t.cache {
+		client.Close()
+	}
+}
+
+// SealConsistencyCheck implement for show msg command line
```
Comment (venkat1109) [inline tools/common/lib.go]: fix comment

File: tools/common/lib.go [hunk 3]
```diff
@@ -1079,3 +1080,148 @@ func getIfSetString(c *cli.Context, p string, setCount *int) (r *string) {
...
+func SealConsistencyCheck(c *cli.Context, mClient mcli.Client) {
+
+	/*
```
Comment (venkat1109) [inline tools/common/lib.go]: remove dead code

File: tools/common/lib.go [hunk 4]
```diff
@@ -1079,3 +1080,148 @@ func getIfSetString(c *cli.Context, p string, setCount *int) (r *string) {
...
+}
+
+func (t *storeClientCache) get(storeUUID string) (*storehost.StoreClientImpl, error) {
+
+	client, ok := t.cache[storeUUID]
+
+	if !ok {
+		hostAddr, err := t.mClient.UUIDToHostAddr(storeUUID)
+		if err != nil {
+			hostAddr = storeUUID + UnknownUUID
```
Comment (venkat1109) [inline tools/common/lib.go]: what does this line do ?

File: tools/common/lib.go [hunk 5]
```diff
@@ -1079,3 +1080,148 @@ func getIfSetString(c *cli.Context, p string, setCount *int) (r *string) {
...
+}
+
+func (t *storeClientCache) close() {
+	for _, client := range t.cache {
+		client.Close()
+	}
+}
+
+// SealConsistencyCheck implement for show msg command line
+func SealConsistencyCheck(c *cli.Context, mClient mcli.Client) {
```
Comment (venkat1109) [inline tools/common/lib.go]: how about moving the scc into its own file ?

File: clients/storehost/client.go [hunk 6]
```diff
@@ -67,3 +67,19 @@ func (s *StoreClientImpl) ReadMessages(req *store.ReadMessagesRequest) (*store.R
...
+func (s *StoreClientImpl) SealExtent(req *store.SealExtentRequest) error {
```
Comment (venkat1109) [inline clients/storehost/client.go]: didn't see this API used, remove if not needed.

File: tools/common/lib.go [hunk 7]
```diff
@@ -1079,3 +1080,188 @@ func getIfSetString(c *cli.Context, p string, setCount *int) (r *string) {
...
+								// now seal the extent
+								if seal {
+
+									fmt.Printf("sealing extent on replica: ", destUUID, extentUUID, storeUUID)
+
+									req := store.NewSealExtentRequest()
+									req.ExtentUUID = common.StringPtr(string(extentUUID))
+									req.SequenceNumber = nil // seal at 'unspecified' seqnum
+
+									// query storage to find address of the message with the given timestamp
```
Comment (datoug) [inline tools/common/lib.go]: update comment

File: tools/common/lib.go [hunk 8]
```diff
@@ -1079,3 +1080,188 @@ func getIfSetString(c *cli.Context, p string, setCount *int) (r *string) {
...
+								fmt.Printf("dest=%v extent=%v store=%v: ", destUUID, extentUUID, storeUUID)
+
+								if extentNotFound {
+									fmt.Printf("not sealed (extent missing)\n", destUUID, extentUUID, storeUUID)
+								} else {
+									fmt.Printf("not sealed\n", destUUID, extentUUID, storeUUID)
+								}
+
+								// now seal the extent
+								if seal {
```
Comment (datoug) [inline tools/common/lib.go]: so we want to send seal request even if the extent is not found?

File: tools/common/lib.go [hunk 9]
```diff
@@ -1079,3 +1080,188 @@ func getIfSetString(c *cli.Context, p string, setCount *int) (r *string) {
...
+
+		for _, desc := range resp.GetDestinations() {
+
+			destUUID := desc.GetDestinationUUID()
+
+			// fmt.Printf("destUUID=%v\n", destUUID)
+
+			listExtentsStats := &shared.ListExtentsStatsRequest{
+				DestinationUUID: common.StringPtr(string(destUUID)),
+				Limit:           common.Int64Ptr(DefaultPageSize),
```
Comment (datoug) [inline tools/common/lib.go]: you can filtered out remote extent by setting localExtentsOnly=true

File: clients/storehost/client.go (hunk 1 above)
Comment (KiranRG) [inline clients/storehost/client.go]: done.

File: clients/storehost/client.go (hunk 6 above)
Comment (KiranRG) [inline clients/storehost/client.go]: it is used now.

File: tools/common/lib.go (hunk 4 above)
Comment (KiranRG) [inline tools/common/lib.go]: i am not 100% sure .. but that seems to be the pattern we have used throughout the code. my sense is it is to ensure we print it out the error with the UUID, etc.

File: tools/common/lib.go (hunk 9 above)
Comment (KiranRG) [inline tools/common/lib.go]: thanks for that; done.

File: tools/common/lib.go (hunk 8 above)
Comment (KiranRG) [inline tools/common/lib.go]: yes; there are scenarios where a store does not even know about an extent .. but it has already been sealed. sending a seal request to a store for such an extent, creates that extent and seals it.

File: tools/common/lib.go (hunk 7 above)
Comment (KiranRG) [inline tools/common/lib.go]: done.

**Pull request merged**